File: 6568.c
#include <stdio.h>
#include <string.h>
char S[9];
int mem[32];
int pc, acc; // pc(5bit/32), acc(8bit/256)

// 오늘은 내가 심심함
int main() {
	while (1) {
		memset(mem, 0, 128);
		for (int i = 0; i < 32; i++) {
			if (!(~scanf("%s ", S))) return 0; // EOF
			for (int j = 0; j < 8; j++) {
				mem[i] = mem[i] * 2 + S[j]-48;
			}
		}

		pc = acc = 0;
		while (1) {
			int op = mem[pc] / 32; // operator
			int val = mem[pc] % 32; // operand
			pc=(pc+1)%32;

			if (op == 7) // HLT
				break;

			switch (op) {
			case 0: // STA x
				mem[val] = acc;
				break;
			case 1: // LDA x
				acc = mem[val];
				break;
			case 2: // BEQ x
				if (!acc)pc = val;
				break;
			case 3: // NOP
				break;
			case 4: // DEC
				acc = (acc + 255) % 256;
				break;
			case 5: // INC
				acc = (acc + 1) % 256;
				break;
			case 6: // JMP x
				pc = val;
				break;
			}
		}

		for (int i = 7; i >= 0; i--)
			printf("%d", (acc >> i) & 1); // & 0000 0001
		puts("");
	}
}

File: Rust/6568.rs
use std::io::{self, Read, Write};

const OPERAND_MASK: u8 = 0b0001_1111;

// 🎅🎄

fn main() {
    let mut input = String::new();
    io::stdin().read_to_string(&mut input).unwrap();
    let mut iter = input.split_whitespace();
    let mut out = io::BufWriter::new(io::stdout().lock());

    loop {
        let mut mem = [0u8; 32];

        for i in 0..32 {
            let s = match iter.next() {
                Some(s) => s,
                None => return,
            };

            let bytes = s.as_bytes();
            for &c in bytes {
                mem[i] = mem[i] * 2 + (c - b'0');
            }
        }

        let mut pc = 0usize;
        let mut acc = 0u8;

        loop {
            let opcode = mem[pc] >> 5;
            let operand = (mem[pc] & OPERAND_MASK) as usize;

            pc = (pc + 1) % 32;

            match opcode {
                0 => { // STA
                    mem[operand] = acc;
                }
                1 => { // LDA
                    acc = mem[operand];
                }
                2 => { // BEQ
                    if acc == 0 {
                        pc = operand;
                    }
                }
                3 => { // NOP }
                4 => { // DEC
                    acc = acc.wrapping_sub(1);
                }
                5 => { // INC
                    acc = acc.wrapping_add(1);
                }
                6 => { // JMP
                    pc = operand;
                }
                7 => { // HLT
                    break;
                }
                _ => unreachable!(),
            }
        }

        writeln!(out, "{:08b}", acc).unwrap();
    }
}
